File: .claude/skills/agent-skill-kit-commons/packages/mjs/src/optimize/rewrite.mjs
import { loadAssets } from "../lint/scorer.mjs";
import { extractVerbsAndNouns } from "./generate.mjs";

function stripVaguePhrases(description, vague) {
  let out = description;
  for (const v of vague) {
    const re = new RegExp(`\\b${v.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")}\\b`, "gi");
    out = out.replace(re, "").replace(/\s+/g, " ").trim();
  }
  return out;
}

function leadsWithVerb(description, verbs) {
  const first = (description ?? "").trim().split(/\W+/)[0]?.toLowerCase();
  return first && verbs.has(first);
}

function pickAnchorVerb(parsed, verbs) {
  const { verbs: candidateVerbs } = extractVerbsAndNouns(parsed);
  for (const v of candidateVerbs) {
    if (verbs.has(v)) return v;
  }
  for (const v of ["scaffold", "validate", "audit", "score", "emit", "resolve", "lint", "optimize", "generate"]) {
    if (verbs.has(v)) return v;
  }
  return null;
}

function ensureTriggerPhrase(description, parsed) {
  if (/\b(use when|when |if |for )/i.test(description)) return description;
  const { nouns } = extractVerbsAndNouns(parsed);
  const noun = nouns[0] ?? "skill";
  return `${description.replace(/\s+$/, "")} Use when the user asks to work with the ${noun}.`;
}

function clampLength(description, low = 60, high = 300) {
  if (description.length >= low && description.length <= high) return description;
  if (description.length > high) {
    const cut = description.slice(0, high - 1);
    const lastSpace = cut.lastIndexOf(" ");
    return (lastSpace > 0 ? cut.slice(0, lastSpace) : cut).replace(/[,;:.]+$/, "") + ".";
  }
  return description;
}

function capitalize(s) {
  if (!s) return s;
  return s[0].toUpperCase() + s.slice(1);
}

/**
 * Three deterministic rewrites. Each emphasizes a different defect class so
 * the user sees varied options. Mutations are stable across runs.
 */
export function proposeVariantsDeterministic(parsed) {
  const { vague, verbs } = loadAssets();
  const base = (parsed.description ?? "").trim();

  // Variant 0: aggressive verb-prefix + trigger phrase
  let v0 = base;
  const anchor = pickAnchorVerb(parsed, verbs);
  if (anchor && !leadsWithVerb(v0, verbs)) {
    v0 = `${capitalize(anchor)} ${v0.replace(/^[A-Z]/, (c) => c.toLowerCase())}`;
  }
  v0 = ensureTriggerPhrase(v0, parsed);
  v0 = clampLength(v0);

  // Variant 1: vague-phrase scrub + trigger phrase
  let v1 = stripVaguePhrases(base, vague);
  if (anchor && !leadsWithVerb(v1, verbs)) {
    v1 = `${capitalize(anchor)} ${v1.replace(/^[A-Z]/, (c) => c.toLowerCase())}`;
  }
  v1 = ensureTriggerPhrase(v1, parsed);
  v1 = clampLength(v1);

  // Variant 2: length-tightening pass (length-first, then verb + trigger)
  let v2 = clampLength(stripVaguePhrases(base, vague));
  if (anchor && !leadsWithVerb(v2, verbs)) {
    v2 = `${capitalize(anchor)} ${v2.replace(/^[A-Z]/, (c) => c.toLowerCase())}`;
  }
  v2 = ensureTriggerPhrase(v2, parsed);
  v2 = clampLength(v2);

  return [
    { text: v0, rationale: "lead-with-verb + append trigger phrase" },
    { text: v1, rationale: "strip vague phrases + lead-with-verb + trigger" },
    { text: v2, rationale: "length-tighten + scrub + verb + trigger" },
  ];
}

File: .claude/skills/agent-skill-kit-commons/packages/py/src/skills_ref/optimize/rewrite.py
from __future__ import annotations

import re

from ..lint.scorer import load_assets
from .generate import extract_verbs_and_nouns


def _strip_vague_phrases(description: str, vague: list[str]) -> str:
    out = description
    for v in vague:
        pat = r"\b" + re.escape(v) + r"\b"
        out = re.sub(pat, "", out, flags=re.IGNORECASE)
        out = re.sub(r"\s+", " ", out).strip()
    return out


def _leads_with_verb(description: str, verbs: set) -> bool:
    first = re.split(r"\W+", (description or "").strip())[0].lower() if description else ""
    return bool(first) and first in verbs


def _pick_anchor_verb(parsed, verbs: set):
    candidate_verbs, _ = extract_verbs_and_nouns(parsed)
    for v in candidate_verbs:
        if v in verbs:
            return v
    for v in [
        "scaffold",
        "validate",
        "audit",
        "score",
        "emit",
        "resolve",
        "lint",
        "optimize",
        "generate",
    ]:
        if v in verbs:
            return v
    return None


def _ensure_trigger_phrase(description: str, parsed) -> str:
    if re.search(r"\b(use when|when |if |for )", description, flags=re.IGNORECASE):
        return description
    _, nouns = extract_verbs_and_nouns(parsed)
    noun = nouns[0] if nouns else "skill"
    return description.rstrip() + f" Use when the user asks to work with the {noun}."


def _clamp_length(description: str, low: int = 60, high: int = 300) -> str:
    if low <= len(description) <= high:
        return description
    if len(description) > high:
        cut = description[: high - 1]
        last_space = cut.rfind(" ")
        trimmed = cut[:last_space] if last_space > 0 else cut
        trimmed = re.sub(r"[,;:.]+$", "", trimmed)
        return trimmed + "."
    return description


def _capitalize(s: str) -> str:
    if not s:
        return s
    return s[0].upper() + s[1:]


def propose_variants_deterministic(parsed) -> list[dict]:
    assets = load_assets()
    vague = assets["vague"]
    verbs = assets["verbs"]
    base = (parsed.description or "").strip()
    anchor = _pick_anchor_verb(parsed, verbs)

    # Variant 0
    v0 = base
    if anchor and not _leads_with_verb(v0, verbs):
        v0 = _capitalize(anchor) + " " + (v0[0].lower() + v0[1:] if v0 else "")
    v0 = _ensure_trigger_phrase(v0, parsed)
    v0 = _clamp_length(v0)

    # Variant 1
    v1 = _strip_vague_phrases(base, vague)
    if anchor and not _leads_with_verb(v1, verbs):
        v1 = _capitalize(anchor) + " " + (v1[0].lower() + v1[1:] if v1 else "")
    v1 = _ensure_trigger_phrase(v1, parsed)
    v1 = _clamp_length(v1)

    # Variant 2
    v2 = _clamp_length(_strip_vague_phrases(base, vague))
    if anchor and not _leads_with_verb(v2, verbs):
        v2 = _capitalize(anchor) + " " + (v2[0].lower() + v2[1:] if v2 else "")
    v2 = _ensure_trigger_phrase(v2, parsed)
    v2 = _clamp_length(v2)

    return [
        {"text": v0, "rationale": "lead-with-verb + append trigger phrase"},
        {"text": v1, "rationale": "strip vague phrases + lead-with-verb + trigger"},
        {"text": v2, "rationale": "length-tighten + scrub + verb + trigger"},
    ]
